src/lib/homestore_backend/hs_hmobj_cp.hpp

```diff
@@ -0,0 +1,70 @@
+/*********************************************************************************
+ * Modifications Copyright 2017-2019 eBay Inc.
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ *********************************************************************************/
+#pragma once
+#include <homestore/checkpoint/cp_mgr.hpp>
+#include <homestore/checkpoint/cp.hpp>
+#include <homestore/homestore_decl.hpp>
+#include <folly/concurrency/ConcurrentHashMap.h>
+#include <homeobject/common.hpp>
+
+#include "hs_homeobject.hpp"
+
+using homestore::CPCallbacks;
+using homestore::CPContext;
+namespace homestore {
+class CP;
+}; // namespace homestore
+
+using homestore::CP;
+
+namespace homeobject {
+
+class HSHomeObject;
+
+class HomeObjCPCallbacks : public CPCallbacks {
+public:
+    HomeObjCPCallbacks(HSHomeObject* home_obj_ptr) : home_obj_(home_obj_ptr){};
+    virtual ~HomeObjCPCallbacks() = default;
+
+public:
+    std::unique_ptr< CPContext > on_switchover_cp(CP* cur_cp, CP* new_cp) override;
+    folly::Future< bool > cp_flush(CP* cp) override;
+    void cp_cleanup(CP* cp) override;
+    int cp_progress_percent() override;
+
+private:
+    HSHomeObject* home_obj_{nullptr}; // it is a raw pointer because HSHomeObject triggers shutdown in its destructor,
```

Comment:
Do we know that the lifecycle of HSHomeObject is greater than HomeObjCPCallbacks?

Reply:
yes. HomeObject in its destruct will call will call homestore shutdown which will do a sync trigger cp which will in return access this homeobj instance. During homestore shutdown, cp manager will be stopped and destroyed which in return destroy all the registered customer of cp manager. After homestore shutdown completes, homeobj destructor continues its destruction.